src/java.base/share/classes/java/lang/StringFormat.java

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (c) 2024, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.lang;
+
+import jdk.internal.math.DoubleToDecimal;
+import jdk.internal.util.HexDigits;
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Formatter;
+import java.util.IllegalFormatConversionException;
+import java.util.function.LongToIntFunction;
+import java.util.function.ToIntFunction;
+
+/**
+ * Utility class for string format fastpath
+ */
+final class StringFormat {
+    static final char DECIMAL_INTEGER           = 'd';
+    static final char HEXADECIMAL_INTEGER       = 'x';
+    static final char HEXADECIMAL_INTEGER_UPPER = 'X';
+    static final char STRING                    = 's';
+
+    static String format(String format, Object... args) {
+        if (args != null) {
+            String s = null;
+            if (args.length == 1) {
+                s = format1(format, args[0]);
+            } else if (args.length == 2) {
+                s = format2(format, args[0], args[1]);
+            }
+            if (s != null) {
+                return s;
+            }
+        }
+
+        return new Formatter().format(format, args).toString();
+    }
+
+    private static String format1(String format, Object arg) {
+        int max = format.length();
+        int off = format.indexOf('%');
+        if (off == -1 || off + 1 == max) {
+            return null;
+        }
+
+        if (off + 2 < max) {
+            if (format.indexOf('%', off + 2) != -1) {
+                return null;
+            }
+        }
+
+        char conv = format.charAt(off + 1);
+        int width = 0;
+        if (conv >= '1' && conv <= '9') {
+            width = conv - '0';
+            if (off + 2 < max) {
+                conv = format.charAt(off + 2);
+            }
+        }
```

Comment:
is it worth handling `width > 9`?

```suggestion
        for (int i = 2; conv >= '1' && conv <= '9' && off + i < max; i++) {
            width = (width * 10) + (conv - '0');
            if (off + i < max) {
                conv = format.charAt(off + i);
            }
        }
```

Reply:
a quick local benchmark on 2021 MacBook M1 Pro before/after motivating this:

```
    @Benchmark
    public String wideStringIntFormat() {
        return "%123s %42d".formatted(s, i);
    }
```

#### Before:
```
Benchmark                          Mode  Cnt    Score    Error  Units
StringFormat.wideStringIntFormat   avgt   15  377.603 ± 85.742  ns/op
```

#### After:
```
Benchmark                          Mode  Cnt    Score    Error  Units
StringFormat.wideStringIntFormat   avgt   15   54.104 ±  0.079  ns/op
```

Reply:
But this is a fastpath optimization, is width > 9 common enough?